tezedge_state/src/proposer/proposal_persister.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) SimpleStaking, Viable Systems and Tezedge Contributors
+// SPDX-License-Identifier: MIT
+
+use std::fs::File;
+use std::io::prelude::*;
+use std::io::BufWriter;
+use std::sync::mpsc;
+
+use crate::proposals::*;
+
+pub struct ProposalPersisterHandle {
+    sender: Option<mpsc::SyncSender<RecordedProposal>>,
+    join_handle: Option<std::thread::JoinHandle<()>>,
+}
+
+impl Drop for ProposalPersisterHandle {
+    fn drop(&mut self) {
+        drop(self.sender.take());
+        if let Some(jh) = self.join_handle.take() {
+            jh.join().unwrap();
+        }
+    }
+}
+
+impl Clone for ProposalPersisterHandle {
+    fn clone(&self) -> Self {
+        // TODO
+        unimplemented!("this is temporary, this type shouldn't be clonable")
+    }
+}
+
+impl ProposalPersisterHandle {
+    pub fn persist<P>(&mut self, proposal: P)
+    where
+        P: Into<RecordedProposal>,
+    {
+        // TODO: better error handling.
+        self.sender
+            .as_mut()
+            .unwrap()
+            .send(proposal.into())
+            .expect("recorded proposal persister disconnected!");
```

Comment:
@binier 
I guess, persister should be used when node is running, so here cannot be `expect`,
this TODO needs to be fixed

Reply:
I guess it comes down to how we should handle errors from which we can't recover. Maybe we should just log error and move on without persister, but idk. If for some reason, if persister failed, that means that those proposals are gone.

Reply:
@binier @jurajselep 
I guess, this is easy, if storing of proposal (for replay) fails here, node cannot crash, and node has to still work:
- bakers should still bake
- indexers should still index data
- node should still apply new blocks
- node should still do a business

node can not crash on "any error", maybe some errors related to business, maybe, if automatic protocol upgrade fails or "disk is full", but I dont see any other relevant error,

if storing of proposal for replay fails, ok, we loose possibility for correct replay, but anyway you can log the whole proposal to log, and if somebody will send us `store-proposals-for-replay-log.data` + `slog.log`, we could still reproduce "correct" replay.

but I am almost sure, that node can not crash on storing proposal here, how would you explain that to baker, that he lost money, because we cannot store proposal? :)

so, back to the topic, so as I wrote in other comments, we better should not use `expect/unwrap/panic` anywhere in code (ok, in test ok), I know we use it on startup, which is also no good, but it will be a good strict rule not to use anywhere just in case

Reply:
@bkontur 
Ok substituted it with logging, so if error happens we log the error and proposal with it. But those proposals will flood the logs and make everything slower and eventually logs will grow so large that we will be out of space.

Maybe we should log here only once and then stop recording proposals, as it's not as important as keeping the node going.